2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,242 @@
+
+`Install-PSResource` will only work for modules.
+A `-AllowPrerelease` switch allows installing prerelease versions.
+Other types will use `Save-PSResource` (see below).
+A `-Repository` parameter accepts a specific repository name.
+If `-Repository` is not specified and the same resource is found in multiple
+repositories, then the resource is automatically installed quietly from the
+trusted repository with the highest version matching the `-Version` parameter
+(if specified, otherwise newest non-prerelease version unless `-AllowPrerelease`
+is used).
+If there are no trusted repositories matching the query, then the newest version
+fulfilling the query will be prompted to be installed.
+If there are multiple repositories with the same trust level containing the same
+version, the first one is used.
+
+`-AllowUntrusted` can be used to suppress being prompted for untrusted sources.
+`-AllowDifferentPublisher` can be used to suppress being prompted if the publisher
+of the module is different from the currently installed version.
+`-AllowReinstall` can be used to re-install a module.
+
+A `-Quiet` switch will suppress progress information.
+
+### Dependencies and version management
+
+`Install-PSResource` can accept a path to a psd1 file (using `-RequiredModulesFile`)
+or a hashtable (using `-RequiredModules`) where the key is the module name and the
+value is either the required version specified using Nuget version range syntax or
+a hash table where `repository` is set to the URL of the repository and
+`version` contains the [Nuget version range syntax](https://docs.microsoft.com/en-us/nuget/reference/package-versioning#version-ranges-and-wildcards).
+
+```powershell
+Install-PSResource -RequiredModules @{
+  "Configuration" = "[1.3.1,2.0)"
+  "Pester"        = @{
+    version = "[4.4.2,4.7.0]"
+    repository = "https://www.powershellgallery.com"
+  }
+}
+```
+
+### Saving resources
+
+With the removal of PackageManagement, there is still a need to support saving
+arbitrary nupkgs (assemblies) used for scripts.
+
+`Save-PSResource -Type Library` will download nupkgs that have a `lib` folder.
+A `-AllowPrerelease` switch allows saving prerelease versions.
+The dependent native library in `runtimes` matching the current system runtime
+will be copied to the same destination.
+
+This cmdlet has the same parameters as `Install-PSResource` with the addition
+of mandatory `-Path` or `-LiteralPath` parameters.
+
+### Updating resources
+
+`Update-PSResource` will update all resources to most recent minor version by
+default.
+A `-OnlyMinorUpdates` switch will allow only updating to newer minor version.
+If the installed resource is a pre-release, it will automatically update to
+latest prerelease or stable version (if available).
+
+### Publishing resources
+
+`Publish-PSResource` will supporting publishing modules and scripts.
+Modules will be given as a path to the module folder.
+Publishing by name and version is no longer supported.
```

Comment:
I have a heavy requirement for package-once-publish-many. `Publish-PSResource` _must_ support publishing the nupkg to a local folder without requiring the existence of the dependencies. Currently, with `publish-module`, you can do this but with a ton of pain. You have to create a folder, register it as a repository, download the nupkgs for the entire dependency graph to that folder, then you can publish the nupkg locally. This process is fragile and takes way to long for what amounts to a fancy call to `nuget pack`

Reply:
For guidance on what I mean, please see my project here https://github.com/markekraus/PSPublishHelper

Reply:
Maybe we should investigate the feasibility of `Publish-PSResource -LocalPublishPath ./path`, where it does all that stuff for you?

Reply:
(Possibly beyond the scope of this RFC, but worth keeping open)

Reply:
If you add an `-OutputPath` to `Publish-PSResource` are we also going to make it take package files as `-Path` so we can publish them to the internet? Because **_that_** is the ask here. Not just **packaging** to a local folder, but also **publishing** that file without needing to re-package it over and over.

The idiomatic PowerShell solution is to have **two commands**, splitting out the packaging part into a separate command, so we write:

```posh
Package-PSResource .\ModuleFolder -To ~\Desktop | Publish-PSResource PSGallery
```

You could still have an overload parameter set on Publish-PSResource that takes a folder and calls `Package` for the user and deletes the artifact ... 

You could even make the `-To` (alias for `-OutputPath` or `-LocalPublishPath` or whatever) parameter optional, and package to temp ... 

I think it's worth it because we want the intermediate artifact for testing and we really shouldn't have to redo the packaging when we inevitably need to publish to multiple feeds.

Reply:
Will support `-DestinationPath` and also `Publish-PSResource` can accept a nupkg.